### ISSUE TITLE
VM only: Floats in between 0 and 1 will evaluate as both == true and == false at the same time if stored in a variable, but will evaluate as != true and != false if used directly

### ISSUE THREAD

**Issue by KormexGit** [YoYoGames/GameMaker-Bugs#2568]
### Description
Doing truthy evaluations with if (value) or if (!value) works as expected, with above 0.5 being true and below 0.5 being false. However, if you instead do if (value == true) or if (value == false), things get weird.
If a number is in between 0 and 1, it will evaluate as being neither true nor false if directly compared to true and false. However, if the exact same value is first stored in a variable, it will evaluate as both true and false at the exact same time.

This weirdness ONLY happens in VM. YYC works as expected, with true and false being treated as 1 and 0, and thus being compared to anything other than 1 and 0 comes back false.

### Steps To Reproduce
```
truthy = 0.1;
if truthy == true {
	show_debug_message("0.1 variable == true"); //appears
}
if truthy == false {
	show_debug_message("0.1 variable == false"); //appears
}

if 0.1 == true {
	show_debug_message("0.1 == true"); //doesn't appear
}
if 0.1 == false {
	show_debug_message("0.1 == false"); //doesn't appear
}
```
On YYC, none of these appear.

[675a1151-ce15-43ee-8212-b493499c0f5a](https://api.gamemaker.io/api/github/downloads/675a1151-ce15-43ee-8212-b493499c0f5a)

**Comment by DragoniteSpam** [comment]
this is possibly the most glorious bug i've ever heard of

**Comment by Alphish** [comment]
Going to paste my bug report from pre-GitHub days, too:

It seems that whenever a variable is compared against a boolean constant, some special comparison is used that is inconsistent with other comparisons. Consider this example:

```
var truthy = 0.7;
var actually_true = true;
var actually_false = false;
show_debug_message("truthy is true: {0}", truthy ? "True" : "False"); // prints True
show_debug_message("truthy == true: {0}", truthy == true ? "True" : "False"); // prints True
show_debug_message("truthy == false: {0}", truthy == false ? "True" : "False"); // prints True
show_debug_message("truthy == actually_true: {0}", truthy == actually_true ? "True" : "False"); // prints False
show_debug_message("truthy == actually_false: {0}", truthy == actually_false ? "True" : "False"); // prints False
show_debug_message("0.7 == true: {0}", 0.7 == true ? "True" : "False"); // prints False
show_debug_message("0.7 == false: {0}", 0.7 == false ? "True" : "False"); // prints False
```

If "truthy" numeric variable is compared against true constant, it returns "True" for any value between 0 and 2 excluded.
If "truthy" numeric variable is compared against false constant, it returns "False" for any value between -1 and 1 excluded.
It also means any numeric variable between 0 and 1 excluded will return "True" when compared against true and false constants.

If truthy numeric variable is compared against a boolean variable, then the variable holding "true" will be equal only to "1", and the variable holding "false" will be equal only to "0", which is much stricter than the 0 : 2 range for true and -1 : 1 range for false.

Likewise, if a numeric constant is compared against a boolean constant, then "true" constant will only equal "1", and the "false" constant will only equal "0".

Both the numeric variable vs bool variable comparison and numeric constant vs bool constant comparison are consistent with one another, while the numeric variable vs bool constant is not. Please make the last comparison consistent with the other two.

**Comment by rwkay** [comment]
I fixed a VM issue earlier in 2024.2 and when I checked this with the current 2024.2 build the VM output matches the YYC output.... closing this

**Comment by Emc1923** [comment]
Verified in Beta IDE v2024.200.0.475 Beta Runtime v2024.200.0.490
